### PR TITLE
Adding support for STR genotyping with HipSTR

### PR DESCRIPTION
**Pull request by nh13**

### CODE REVIEW

**Comment by nh13** [comment]
@tfenne I would like a "lite" review of this PR, in terms of direction and major objections (your call, it's subjective), as well as recommendations for future development.  I am going to use the `devel` branch for all of this R&D work, as we currently don't know if this is going to be merged into the stable branch until we have tested this on many more samples.

Could you also take a look at the potential [README](https://github.com/fulcrumgenomics/fgstr/tree/nh_hipstr_1) that will be viewed on the home page to see if it makes sense to you?

File: README.md [hunk 1]
```diff
@@ -1 +1,60 @@
 # fgstr
+
+Simple Tandem Repeat for Duplex Sequencing
```
Comment (tfenne) [inline README.md]: Perhaps "Simple Tandem Repeat Genotyping for Duplex Sequencing"

File: README.md [hunk 2]
```diff
@@ -1 +1,60 @@
...
+
+## *** WARNING ***
+
+This tools and pipelines in this repository are in active development, and are not intended for production use.
+They are subject to change at any time as we develop new methods and procedure
```
Comment (tfenne) [inline README.md]: procedure[s]

File: README.md [hunk 3]
```diff
@@ -1 +1,60 @@
...
+
+
+## STR Calling a Single Sample with Duplex Sequencing
+
+To analyze a single sample:
+
+1. Run `PrepareUnmappedBamPipeline` to create an unmapped BAM with extracted UMI information.
+2. Run `MapAndGroupRawReads` to map the raw reads and assign them each to a signle source duplex molecule.
```
Comment (tfenne) [inline README.md]: signle -> single

File: README.md [hunk 4]
```diff
@@ -1 +1,60 @@
...
+3. Run `GenotypeFromGroupedBam` to ssign reads from the same source molecule to their own read group, filter the reads from the same source
```
Comment (tfenne) [inline README.md]: ssign -> assign

File: README.md [hunk 5]
```diff
@@ -1 +1,60 @@
...
+Simple Tandem Repeat for Duplex Sequencing
+
+## *** WARNING ***
+
+This tools and pipelines in this repository are in active development, and are not intended for production use.
+They are subject to change at any time as we develop new methods and procedure
+
+
+## STR Calling a Single Sample with Duplex Sequencing
+
```
Comment (tfenne) [inline README.md]: I'd add a small paragraph here saying that the current approach is to use HipSTR and treat each duplex molecule as a single haploid sample, etc.

File: README.md (hunk 4 above)
Comment (tfenne) [reply README.md]: This paragraph is hard to parse as it's a bit of a run on sentence.  Perhaps change it to:

Run `GenotypeFromGroupedBam` to generate STR genotypes from the grouped BAM.  This entails a) generating a BAM with a read-group per source molecule, b) running HipSTR to jointly call across all source molecules (treating each source as a single haploid sample), and c) combining the per-molecule "genotypes" to produce a final genotype.

File: README.md [hunk 6]
```diff
@@ -1 +1,60 @@
...
+genotypes jointly across all source duplex molecules (treats each source molecule as its own sample), and
+combine the per-source-duplex-molecule genotypes into a final genotype.
+
+### Input Requirements
+
+The interval list provided to `GenotypeFromGroupedBam` requires a custom format.
+The name field should contain a comma list of values as follows:
+
+  1. the repeat unit length (ex. `3` for the tri-nucleotide repeat `TCATCATCATCA`).
+  2. the number of repeat units (ex. `4` for the tri-nucleotide repeat `TCATCATCATCA`).
```
Comment (tfenne) [inline README.md]: Number of repeat units where?  In the reference genome?  Or can I have a value here that's different from what's in the ref.fasta?

File: README.md [hunk 7]
```diff
@@ -1 +1,60 @@
...
+  3. the name of the STR (ex. `D1S1656`)
+  4. optionally, the expected (known or truth) number of repeat units for allele #1
```
Comment (tfenne) [inline README.md]: If these are optional, how do I omit them?  Is it `4,17,D1S1656,,` or `4,17,D1S1656`? Or something else?

File: README.md [hunk 8]
```diff
@@ -1 +1,60 @@
...
+
+Simple Tandem Repeat for Duplex Sequencing
+
+## *** WARNING ***
+
+This tools and pipelines in this repository are in active development, and are not intended for production use.
+They are subject to change at any time as we develop new methods and procedure
+
+
+## STR Calling a Single Sample with Duplex Sequencing
```
Comment (tfenne) [inline README.md]: Before this, we probably need a short section on building the tools.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/DuplexConsensusPipeline.scala [hunk 9]
```diff
@@ -26,23 +26,24 @@
...
-import dagr.commons.io.Io
+import com.fulcrumgenomics.commons.io.Io
 import dagr.core.cmdline.Pipelines
 import dagr.core.tasksystem.{Pipeline, ShellCommand}
 import dagr.tasks.DagrDef._
 import dagr.tasks.bwa.Bwa
 import dagr.tasks.fgbio._
 import dagr.tasks.picard._
 
+@deprecated(message="See the README for more information: https://github.com/fulcrumgenomics/fgstr", since="Jun-7-2017")
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/DuplexConsensusPipeline.scala]: Why not just yank it out?  Since everything is still a bit experimental and subject to change I'd vote for deleting vs. deprecating.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+  description =
+    """
+      |Pipeline to call STR genotypes from a grouped BAM.
+      |
+      |Assigns reads from the same source molecule to their own read group, filters the reads from the same source
+      |duplex molecule to require a minimum number of reads (overall, AB strand, and BA strand), runs HipSTR to call
+      |genotypes jointly across all source duplex molecules (treats each source molecule as its own sample), and
+      |combines the per-source-duplex-molecule genotypes into a final genotype.
+      |
+      |The grouped BAM should be created by fgbio's `GroupReadsByUmi`
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: ... with --strategy=paired, or by running the <insert name here> pipeline.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+      |
+      |Requires as input the output path prefix given to DuplexConsensusPipeline.
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: Is this no longer true?  `input` is a BAM below, and `DuplexConsensusPipeline` is deprecated.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+      |
+      |An interval list of specifying the same regions as input to HipSTR should be given.  The name field should contain
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: This doesn't make a lot of sense to me.  Is this saying an interval list specifying the regions I would send to HipSTR if I were running HipSTR outside of this pipeline?  Perhaps just "An interval list specifying the set of regions over which to search for STRs should be given."

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+(@arg(flag='i', doc="Input SAM or BAM file (from GroupReadsByUmi).") val input: PathToBam,
+ @arg(flag='r', doc="Path to the reference FASTA.")              val ref: PathToFasta,
+ @arg(flag='l', doc="STR regions to analyze.")                   val intervals: PathToIntervals,
+ @arg(flag='o', doc="Path prefix for output files.")             val output: PathPrefix,
+ @arg(flag='M', minElements=1, maxElements=3, doc="The minimum number of raw reads per source molecule.")
+ val minReads: Seq[Int]
+) extends Pipeline(Some(output.getParent)) {
+
+  override def build(): Unit = {
+
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: bonus newline!

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    def f(ext: String): FilePath = PathUtil.pathTo(output + ext)
+
+    val bed                   = f(".regions.bed")
+    val haploidChromosomeFile = f(".haploid.chromosomes.txt")
+    val readGroupBam          = f(".read_group_per_mid.bam")
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: mid?  
image

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+    val filteredBam           = f(".read_group_per_mid.filtered.bam")
+    val hipStrVcf             = f(".hipstr.vcf.gz")
+    val finalOutput           = f("")
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: This is a little weird.  If `StrGenotypeDuplexMolecules` wants a path prefix, I think I'd just pass `output` directly there instead of doing this.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+    val toBed         = new CreateRegionsBed(intervals=intervals, bed=bed)
+    val toHapFile     = new CreateHaploidChromosomeFile(ref=ref, output=haploidChromosomeFile)
+    val toRgBam       = new ReadGroupPerDuplexMolecularId(input=input, output=readGroupBam, ref=Some(ref), intervals=Some(intervals))
+    val toFilteredBam = new FilterRawReadsFromDuplexMolecule(input=readGroupBam, output=filteredBam, ref=Some(ref), minReads=minReads)
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: I'll probably understand when I get there, but I figured the filtering would just happen in the conversion to rg-per-mi tool

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+    val toHipstrVcf   = new HipStr(input=filteredBam, ref=ref, regions=bed, output=hipStrVcf, haploidPath=Some(haploidChromosomeFile))
+    val toFinalVcf    = new StrGenotypeDuplexMolecules(input=hipStrVcf, output=finalOutput, intervals=intervals)
+
+    root ==> (toBed :: toHapFile) ==> toHipstrVcf
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: This can all be one glorious expression:

```scala
root ==> (toBed :: toHapFile :: (toRgBam ==> toFilteredBam)) ==> toHipstrVcf ==> toFinalVcf
```

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/MapAndGroupRawReads.scala [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+    val mappedRaw   = dir.resolve(pre + ".raw.aligned.bam")
+    val grouped     = dir.resolve(pre + ".grouped.bam")
+    val familySizes = dir.resolve(pre + ".family_sizes.txt")
+
+    // Make the grouped and consensus BAMs
+    val bwaRaw    = Bwa.bwaMemStreamed(unmappedBam=unmappedBam, mappedBam=mappedRaw, ref=ref, samToFastqCores=0, bwaMemMemory="6G")
+    val group     = new GroupReadsByUmi(in=mappedRaw, out=grouped, familySizeOut=Some(familySizes),
+      rawTag=Some(umiTag), minMapQ=Some(minMapQ), strategy=AssignmentStrategy.Paired, edits=Some(edits))
+    val dsMetrics         = new CollectDuplexSeqMetrics(input=grouped, output=dir.resolve(pre), intervals=Some(intervals))
+    val dsMetricsAllReads = new CollectDuplexSeqMetrics(input=grouped, output=dir.resolve(pre + ".all_reads"), intervals=None)
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/MapAndGroupRawReads.scala:77]: I'm curious if you're finding this output useful?

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/PostDuplexConsensusPipeline.scala [hunk 19]
```diff
@@ -28,15 +28,15 @@ package com.fulcrumgenomics.str.pipelines
...
+import com.fulcrumgenomics.commons.io.{Io, PathUtil}
 import dagr.core.cmdline.Pipelines
 import dagr.core.tasksystem.Pipeline
 import dagr.tasks.DagrDef.{PathPrefix, PathToBam, PathToFasta, PathToIntervals}
 import dagr.tasks.fgbio.{ClipOverlappingReads, FilterConsensusReads}
 import dagr.tasks.picard.IntervalListToBed
 import dagr.tasks.vc.VarDictJavaEndToEnd
 
-
+@deprecated(message="See the README for more information: https://github.com/fulcrumgenomics/fgstr", since="Jun-7-2017")
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/PostDuplexConsensusPipeline.scala:39]: Ditto on deleting vs. deprecating.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/tasks/HipStr.scala [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+import com.fulcrumgenomics.commons.CommonsDef._
+import htsjdk.samtools.util.IntervalList
+
+import scala.collection.mutable.ListBuffer
+
+object HipStr extends Configuration {
+  type PathToBed = FilePath
+  val HipStrExecutableConfigKey: String = "hipstr.executable"
+
+  def findHipStr: Path = configureExecutable(HipStrExecutableConfigKey, "HipSTR")
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/tasks/HipStr.scala]: Do you want `configureExecutableFromPath` here?

File: pipelines/src/main/scala/com/fulcrumgenomics/str/tasks/HipStr.scala [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+
+  /** Creates a file with one chromosome name per line. */
+  class CreateHaploidChromosomeFile(ref: PathToFasta, output: FilePath) extends SimpleInJvmTask {
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/tasks/HipStr.scala]: Just a thought that you could also do this by getting the unique list of chromosomes referenced in the interval-list/BED file.  That way you wouldn't end up including the ~3000 or so alt and decoys.  Not sure if that matters or not thought.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/tasks/HipStr.scala [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+      Io.writeLines(path=output, lines=dict.getSequences.map(_.getSequenceName).toSeq)
+    }
+  }
+
+  /** Creates the regions BED file for HipStr: Required format is tab-delimited columns CHROM START STOP PERIOD NCOPIES */
+  class CreateRegionsBed(intervals: PathToIntervals, bed: PathToBed) extends SimpleInJvmTask {
+    def run(): Unit = {
+      val list  = IntervalList.fromFile(intervals.toFile)
+      val lines = list.map { interval =>
+        val nameTokens = interval.getName.split(',')
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/tasks/HipStr.scala:52]: Maybe validate that this is always 3-5 in length?  Just in case a comma sneaks into a name at some point.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/tasks/HipStr.scala [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+class HipStr(input: PathToBam,
+             ref: PathToFasta,
+             regions: PathToBed,
+             output: PathToVcf,
+             genotypeLikelihoods: Boolean = true,
+             posteriorLikelihoods: Boolean = true,
+             minReads: Option[Int] = Some(1),
+             useUnpaired: Boolean = true,
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/tasks/HipStr.scala]: I'm a little wary of this ... just because we wouldn't generally create consensus reads from frags or reads with unmapped mates.

File: tools/src/main/resources/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.R [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    df = data.frame(numUnits, counts)
+    p = ggplot(df) + aes(x=numUnits) +
+        geom_point(aes(y=counts, color=counts)) +
+        scale_x_continuous() +
+        scale_fill_manual(values=fgcolors) +
+        scale_color_gradient(low=fgblue, high=fggreen) +
+        labs(x="# of Repeat Units", y="Raw counts") +
+        theme(plot.title = element_text(hjust = 0.5), legend.title=element_blank()) +
+        ggtitle(paste(strName, " (", chromosome, ":", start, "-", end, ")", sep=""))
```
Comment (tfenne) [inline tools/src/main/resources/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.R]: you can just use `title=...` in `labs()` for this.

File: tools/src/main/resources/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.R [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+pdf(outputFile, width=11, height=8.5)
+
+for (i in 1:nrow(data)) {
+    row = data[i,]
+    chromosome  = row[1]
+    start       = row[2]
+    end         = row[3]
+    unitLength  = row[4]
+    refNumUnits = row[5]
+    strName     = row[[6]]
```
Comment (tfenne) [inline tools/src/main/resources/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.R:53]: Why do you switch to double `[[]]` here?  Is that meaningful?

File: tools/src/main/resources/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.R [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE   #
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER        #
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, #
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN     #
+# THE SOFTWARE.                                                                 #
+#################################################################################
+
+#################################################################################
+# R script to generate QC plots VcfToGenotypes
+#################################################################################
```
Comment (tfenne) [inline tools/src/main/resources/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.R]: Some documentation of args, inputs and outputs would be nice.

File: tools/src/main/scala/com/fulcrumgenomics/str/bam/FilterRawReadsFromDuplexMolecule.scala [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+  val FilterFragments   = "being Fragment/Non-Paired Reads"
+  val FilterFailed      = "did not meet thresholds"
+}
+
+@clp(group=ClpGroups.VcfOrBcf, description=
+  """
+    |Filters raw reads from the same duplex source molecule.
+    |
+    |The raw reads should have been grouped using fgbio's GroupReadsByUmi.
+    |
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/bam/FilterRawReadsFromDuplexMolecule.scala]: Extra newline.

File: tools/src/main/scala/com/fulcrumgenomics/str/bam/FilterRawReadsFromDuplexMolecule.scala [hunk 28]
```diff
@@ -0,0 +1,220 @@
...
+import com.fulcrumgenomics.commons.io.Io
+import com.fulcrumgenomics.commons.util.{LazyLogging, SimpleCounter}
+import com.fulcrumgenomics.sopt.{arg, clp}
+import com.fulcrumgenomics.str.FgStrDef.toMolecularId
+import com.fulcrumgenomics.str.cmdline.FgStrTool
+import com.fulcrumgenomics.umi.ConsensusTags
+
+object FilterRawReadsFromDuplexMolecule {
+  /** Additional filter strings used when rejecting reads. */
+  val FilterFragments   = "being Fragment/Non-Paired Reads"
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/bam/FilterRawReadsFromDuplexMolecule.scala]: Doesn't this always come out as zero? GroupReadsByUmi shouldn't emit non-paired reads or reads without both mates mapped.

File: tools/src/main/scala/com/fulcrumgenomics/str/bam/FilterRawReadsFromDuplexMolecule.scala [hunk 29]
```diff
@@ -0,0 +1,220 @@
...
+  private case class Filters(minReadsPerMolecule: Int, minReadsAb: Int, minReadsBa: Int)
+  private val filters = Filters(
+    minReadsPerMolecule = this.minReads.head,
+    minReadsAb          = this.minReads.take(2).last,
+    minReadsBa          = this.minReads.last
+  )
+  validate(filters.minReadsAb <= filters.minReadsPerMolecule, "min-reads values must be specified high to low.")
+  validate(filters.minReadsBa <= filters.minReadsAb, "min-reads values must be specified high to low.")
+
+  override def execute(): Unit = {
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/bam/FilterRawReadsFromDuplexMolecule.scala]: I think this whole thing could be _much_ simpler by either:
1) Just slurping in the next group of reads, discarding R2s, and then counting a) total size, b) recs ending with /A and c) recs ending with /B.  No need for all the subgrouping into AbR1/AbR2/BaR1/BaR2 etc.
2) Just running the consensus caller on the group and harvesting the stats

(2) would of give a lower bound as the caller itself would discard some reads due to all of it's filtering.  So I would suggest (1) ... and even simplifying it to the point it could be implemented as a simple iterator and then slotted into the tool that makes a RG per MI.

File: tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala [hunk 30]
```diff
@@ -0,0 +1,156 @@
...
+    |appended (also with a dash separator).
+    |
+    |An interval list can be given to filter output reads to a set of intervals.
+  """)
+class ReadGroupPerDuplexMolecularId
+(
+  @arg(flag='i', doc="Input SAM or BAM file (from GroupReadsByUmi).") val input: PathToBam,
+  @arg(flag='o', doc="Output SAM or BAM.") val output: PathToBam = Io.StdOut,
+  @arg(flag='r', doc="Reference sequence fasta file.") val ref: Option[PathToFasta] = None,
+  @arg(flag='l', doc="Optional list of intervals to reads to.") val intervals: Option[PathToIntervals] = None,
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala]: `to reads to`

File: tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala [hunk 31]
```diff
@@ -0,0 +1,156 @@
...
+    logger.info("Extracting molecular ids")
+    val mids = {
+      val progress = ProgressLogger(this.logger, unit=5e5.toInt, verb="Extracted molecule ids")
+      val in       = SamSource(input, ref=ref)
+      val iterator = toIterator(in)
+      val ids      = new mutable.HashSet[Long]()
+      iterator.foreach { rec =>
+        ids += toMolecularId(rec, this.assignTag)
+        progress.record(rec)
+        if (progress.getCount % 5e5.toInt == 0) {
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala]: you can just do `if (progress.record(rec)) ....` since it returns Boolean with true = it printed something and false = it didn't.

File: tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala [hunk 32]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  /** Create the output header. */
+  private def toOutputHeader(in: SamSource, mids: Seq[Long]): SAMFileHeader = {
+    val header = in.header.clone()
+
+    val sourceReadGroup = header.getReadGroups.toList match {
+      case Nil       => throw new IllegalArgumentException(s"Found no read groups in ${this.input}")
+      case rg :: Nil => rg
+      case rgs       => throw new IllegalArgumentException(s"Found more than one read group in ${this.input}")
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala]: I think this is fine for now, but you could presumably handle multiple where the library/sample are the same right?

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala [hunk 33]
```diff
@@ -0,0 +1,101 @@
...
+class GenotypeFromGroupedBam
+(@arg(flag='i', doc="Input SAM or BAM file (from GroupReadsByUmi).") val input: PathToBam,
+ @arg(flag='r', doc="Path to the reference FASTA.")              val ref: PathToFasta,
+ @arg(flag='l', doc="STR regions to analyze.")                   val intervals: PathToIntervals,
+ @arg(flag='o', doc="Path prefix for output files.")             val output: PathPrefix,
+ @arg(flag='M', minElements=1, maxElements=3, doc="The minimum number of raw reads per source molecule.")
+ val minReads: Seq[Int]
+) extends Pipeline(Some(output.getParent)) {
+
+  override def build(): Unit = {
```
Comment (tfenne) [inline pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: Looking back at this pipeline after having gone through some of the tools, I wonder if it wouldn't make more sense to parallelize (scatter gather?) by STR here?  I think that would have a couple of advantages:

1. You'd create multiple BAM, but each would have ~20-fold fewer read groups in them.  I imagine this would make the output VCFs from HipSTR more tractable too since you would have fewer samples, and a lot less no-calls.
2. The pipeline would be a lot more parallelizable, so could complete a lot quicker if >> 1 CPU is available

File: tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala [hunk 34]
```diff
@@ -0,0 +1,357 @@
...
+        s"$genotype:${allele.count}"
+      }.mkString(",")
+      s"$this\t$genotypes"
+    }
+  }
+}
+
+@clp(group=ClpGroups.VcfOrBcf, description=
+  """
+    |Calls a genotype for individually called duplex molecules.
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala]: [diploid] genotype

File: tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala [hunk 35]
```diff
@@ -0,0 +1,357 @@
...
+  """)
+class StrGenotypeDuplexMolecules
+(
+  @arg(flag='i', doc="Input VCF of somatic variant calls.") val input: PathToVcf,
+  @arg(flag='o', doc="Prefix for all output files.") val output: PathPrefix,
+  @arg(flag='l', doc="Interval list with the STR regions.") val intervals: PathToIntervals
+) extends FgStrTool with LazyLogging {
+  import StrGenotypeDuplexMolecules.StrInfo
+
+  // TODO: do we want to output no-calls for missing sites?
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala]: I think that would be good, yes.

File: tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala [hunk 36]
```diff
@@ -0,0 +1,357 @@
...
+            infos.head
+        }
+
+        val ctxBuilder = new VariantContextBuilder(ctx)
+        val genotypes  = ctx.getGenotypes
+        val alleles    = ctx.getAlleles.toSeq
+
+        // get the count per allele
+        val counts: Seq[Int] =  alleles.map { allele => genotypes.map { genotype => genotype.countAllele(allele) }.sum }
+        require(alleles.length == counts.length, s"Found ${alleles.length} alleles and ${counts.length} counts")
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala]: This seems impossible to fail.  Seq.map is guaranteed to produce a Seq of the same length.

File: tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala [hunk 37]
```diff
@@ -0,0 +1,357 @@
...
+        val allCalls = alleles.zip(counts)
+          .filter(_._2 > 0) // ignore zero counts
+          .map { case (allele, count) =>
+            val baseDiff     = allele.length() - refAlleleLength
+            val repeatLength = baseDiff + (strInfo.refLength * strInfo.unitLength)
+            StrAllele(allele=allele, repeatLength=repeatLength, count=count)
+          }
+          .sortBy(-_.count)
+
+        val genotypeCalls = {
```
Comment (tfenne) [inline tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala]: I think this needs to be marginally less naive.  I think you want something like:

1. Take the first allele
2. If the first allele doesn't represent > 80% (?) of the counts, also take the second allele

Otherwise anyone who is homozygous will get called het so long as there is at least one error.

**Review by tfenne — COMMENTED** [review]
I didn't really look at the tests, and I did a fairly light pass (as requested).  Some nit-picky comments just because I saw them.  And a few larger suggestions/comments.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala (hunk 15 above)
Comment (nh13) [reply pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/GenotypeFromGroupedBam.scala]: Doh, it wanted a vcf originally, and I got lazy.

File: pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/MapAndGroupRawReads.scala (hunk 18 above)
Comment (nh13) [reply pipelines/src/main/scala/com/fulcrumgenomics/str/pipelines/MapAndGroupRawReads.scala:77]: I want it there for a large set of samples that should be appearing, and then I'll let you know.

File: tools/src/main/resources/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.R (hunk 25 above)
Comment (nh13) [reply tools/src/main/resources/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.R:53]: So I don't get a list returned in some cases: https://stackoverflow.com/questions/1169456/the-difference-between-and-notations-for-accessing-the-elements-of-a-lis

File: tools/src/main/scala/com/fulcrumgenomics/str/bam/FilterRawReadsFromDuplexMolecule.scala (hunk 28 above)
Comment (nh13) [reply tools/src/main/scala/com/fulcrumgenomics/str/bam/FilterRawReadsFromDuplexMolecule.scala]: I am just going to leave this here: https://github.com/fulcrumgenomics/fgbio/blob/master/src/main/scala/com/fulcrumgenomics/umi/DuplexConsensusCaller.scala#L49

File: tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala (hunk 31 above)
Comment (nh13) [reply tools/src/main/scala/com/fulcrumgenomics/str/bam/ReadGroupPerDuplexMolecularId.scala]: @tfenne I actually can't, since given an fgbio `SamRecord`, it returns `Unit`!

File: tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala (hunk 37 above)
Comment (nh13) [reply tools/src/main/scala/com/fulcrumgenomics/str/vcf/StrGenotypeDuplexMolecules.scala]: Alternatively, it is calling low-frequency mutations!
